vncviewer/BaseTouchHandler.cxx

```diff
@@ -0,0 +1,151 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <stdlib.h>
+#include <math.h>
+
+#define XK_MISCELLANY
+#include <rfb/keysymdef.h>
+
+#include "GestureHandler.h"
+#include "BaseTouchHandler.h"
+
+// Sensitivity threshold for gestures
+static const int ZOOMSENS = 30;
+static const int SCRLSENS = 50;
+
+BaseTouchHandler::BaseTouchHandler()
+{
+}
+
+BaseTouchHandler::~BaseTouchHandler()
+{
+}
+
+void BaseTouchHandler::handleGestureEvent(const GHEvent& ev)
+{
+  double magnitude;
+
+  switch (ev.type) {
+  case GH_GestureBegin:
+    switch (ev.gesture) {
+    case GH_ONETAP:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 1, ev);
+      fakeButtonEvent(false, 1, ev);
+      break;
+    case GH_TWOTAP:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 3, ev);
+      fakeButtonEvent(false, 3, ev);
+      break;
+    case GH_THREETAP:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 2, ev);
+      fakeButtonEvent(false, 2, ev);
+      break;
+    case GH_DRAG:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 1, ev);
+      break;
+    case GH_LONGPRESS:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 3, ev);
+      break;
+    case GH_PAN:
+      lastMagnitudeX = ev.magnitude_x;
+      lastMagnitudeY = ev.magnitude_y;
+      fakeMotionEvent(ev);
+      break;
+    case GH_PINCH:
+      lastMagnitudeX = sqrt(ev.magnitude_x * ev.magnitude_x +
+                            ev.magnitude_y * ev.magnitude_y);
+      fakeMotionEvent(ev);
+      break;
+    }
+    break;
+
+  case GH_GestureUpdate:
+    switch (ev.gesture) {
+    case GH_DRAG:
+    case GH_LONGPRESS:
+      fakeMotionEvent(ev);
+      break;
+    case GH_PAN:
+      while ((ev.magnitude_y - lastMagnitudeY) > SCRLSENS) {
+        fakeButtonEvent(true, 4, ev);
+        fakeButtonEvent(false, 4, ev);
+        lastMagnitudeY += SCRLSENS;
+      }
+      while ((ev.magnitude_y - lastMagnitudeY) < -SCRLSENS) {
+        fakeButtonEvent(true, 5, ev);
+        fakeButtonEvent(false, 5, ev);
+        lastMagnitudeY -= SCRLSENS;
+      }
+      while ((ev.magnitude_x - lastMagnitudeX) > SCRLSENS) {
+        fakeButtonEvent(true, 6, ev);
+        fakeButtonEvent(false, 6, ev);
+        lastMagnitudeX += SCRLSENS;
+      }
+      while ((ev.magnitude_x - lastMagnitudeX) < -SCRLSENS) {
+        fakeButtonEvent(true, 7, ev);
+        fakeButtonEvent(false, 7, ev);
+        lastMagnitudeX -= SCRLSENS;
+      }
+      break;
+    case GH_PINCH:
+      magnitude = sqrt(ev.magnitude_x * ev.magnitude_x +
+                       ev.magnitude_y * ev.magnitude_y);
+      if (abs(magnitude - lastMagnitudeX) > ZOOMSENS) {
```

Comment:
What's the thought behind this?

Magnitude for PINCH is the distance between the fingers (see "GestureHandler.cxx:81). We only act on a GH_PINCH if the difference between magnitude and lastMagnitude is large enough.. So if I do a really slow PINCH, the difference between magnitude and lastMagnitude must be constantly smaller than ZOOMSENS which means that it should be ignored?

Reply:
Ah, lastMagnitude is only set on GH_GestureBegin and not on every event.